wikipedia/_tools/parse_clicks.py

```diff
@@ -0,0 +1,106 @@
+console_handler = logging.StreamHandler()
+console_handler.setLevel(logging.INFO)
+
+# Define a formatter for the log messages
+formatter = logging.Formatter("%(asctime)s - %(levelname)s - %(message)s")
+console_handler.setFormatter(formatter)
+
+# Add the console handler to the logger
+logger.addHandler(console_handler)
+
+
+class ClickStreamDist:
+    def __init__(self, year=2023, month=6, lang="en"):
+        self.filename = f"clickstream-{lang}wiki-{year}-{month:02d}.tsv.gz"
+        self.url = f"https://dumps.wikimedia.org/other/clickstream/{year:d}-{month:02d}/{self.filename}"
+        self.clickstream_output_file = "clickstream.tsv.gz"
+
+    def download(self):
+        if os.path.exists(self.clickstream_output_file):
+            logger.info("File already exists. Skipping download.")
+            return
+
+        logger.info("Downloading the clickstream file...")
+        response = requests.get(self.url)
+
+        if response.status_code == 200:
+            with open(self.clickstream_output_file, "wb") as file:
+                file.write(response.content)
+            logger.info("File downloaded successfully.")
+        else:
+            logger.info("Failed to download the file.")
+
+    def analyze(self):
+        logger.info("Analyzing...")
+
+        word_freq = self.calculate_word_frequency()
+        word_prob = self.calculate_word_probability(word_freq)
+        sorted_prob = self.sort_probability_distribution(word_prob)
+
+        self.dump_probability_distribution(sorted_prob)
+
+        logger.info("Analysis completed.")
+
+    def calculate_word_frequency(self):
+        logger.info("Calculating word frequency...")
+
+        word_freq = Counter()
+        with gzip.open(self.clickstream_output_file, "rt", encoding="utf-8") as file:
+            reader = csv.reader(file, delimiter="\t")
+            for row in reader:
+                if row[0] != "other-search":
+                    word = row[1]
+                    count = int(row[3])
+                    word_freq[word] += count
+        return word_freq
+
+    def calculate_word_probability(self, word_freq):
+        logger.info("Calculating word probability...")
+
+        total_words = sum(word_freq.values())
+        word_prob = {word.replace("_", " "): count / total_words for word, count in word_freq.items()}
+
+        # Delete the 'Main Page' key
+        del word_prob["Main Page"]
+
+        return word_prob
+
+    def sort_probability_distribution(self, word_prob):
+        logger.info("Sorting probability distribution...")
+
+        sorted_prob = sorted(word_prob.items(), key=lambda x: x[1], reverse=True)
+        return sorted_prob
+
+    def dump_probability_distribution(self, sorted_prob):
+        logger.info("Dumping probability distribution...")
+
+        # Dump the first 10000 entries to a different file
+        writer = csv.writer(sys.stdout)
+        writer.writerow(["Word", "Probability"])
+        writer.writerows(sorted_prob[:10000])
```

Comment:
I think the truncate is done too late. The sum of all probabilities is not equals to 1 anymore.

Instead, the queries list should be truncated before the probabilities are computed.

Your algorithm is:
- Compute probabilities
- Filter out value we do not want (Main Page)
- Sort by probabilities
- Keep the 10000 first result

What I think the algorithm should be:
- Filter out value we do not want (Main Page)
- Sort by count and keep only the first 10000 results
- Compute the probabilities of the latter step

Reply:
Updated